test/AssetPool.test.js

```diff
@@ -975,4 +975,278 @@ describe("AssetPool", () => {
       )
     })
   })
+
+  describe("approveNewAssetPoolUpgrade", () => {
+    let newAssetPool
+
+    beforeEach(async () => {
+      const NewUnderwriterToken = await ethers.getContractFactory(
+        "UnderwriterToken"
+      )
+      newUnderwriterToken = await NewUnderwriterToken.deploy(
+        "New Underwriter Token",
+        "newCOV"
+      )
+      await newUnderwriterToken.deployed()
+
+      const NewAssetPoolStub = await ethers.getContractFactory(
+        "NewAssetPoolStub"
+      )
+      newAssetPool = await NewAssetPoolStub.deploy(
+        collateralToken.address,
+        newUnderwriterToken.address
+      )
+      await newAssetPool.deployed()
+    })
+
+    context("when called by the governance", () => {
+      it("should approve new asset pool", async () => {
+        await assetPool
+          .connect(coveragePool)
+          .approveNewAssetPoolUpgrade(newAssetPool.address)
+
+        expect(await assetPool.newAssetPool()).to.equal(newAssetPool.address)
+      })
+
+      it("should emit ApprovedAssetPoolUpgrade event", async () => {
+        const tx = await assetPool
+          .connect(coveragePool)
+          .approveNewAssetPoolUpgrade(newAssetPool.address)
+
+        expect(tx)
+          .to.emit(assetPool, "ApprovedAssetPoolUpgrade")
+          .withArgs(newAssetPool.address)
+      })
+    })
+
+    context("when called not by the governance", () => {
+      it("should revert", async () => {
+        await expect(
+          assetPool
+            .connect(thirdParty)
+            .approveNewAssetPoolUpgrade(newAssetPool.address)
+        ).to.be.revertedWith("Ownable: caller is not the owner")
+      })
+    })
+  })
+
+  describe("upgradeToNewAssetPool", () => {
+    let newAssetPool
+    let newUnderwriterToken
+
+    beforeEach(async () => {
+      const NewUnderwriterToken = await ethers.getContractFactory(
+        "UnderwriterToken"
+      )
+      newUnderwriterToken = await NewUnderwriterToken.deploy(
+        "New Underwriter Token",
+        "newCOV"
+      )
+      await newUnderwriterToken.deployed()
+
+      const NewAssetPool = await ethers.getContractFactory("NewAssetPoolStub")
+      newAssetPool = await NewAssetPool.deploy(
+        collateralToken.address,
+        newUnderwriterToken.address
+      )
+      await newAssetPool.deployed()
+      await newUnderwriterToken.transferOwnership(newAssetPool.address)
+    })
+
+    context(
+      "when a governance was compromised and it approved a fake new asset pool",
```

Comment:
Can we say instead "when new asset pool address does not match"? This could be a user error as well. Or, in extreme circumstances, upgrade of an upgrade conflicting with a pending underwriter's transaction.

Reply:
Actually... we can leave it as is, after reading the test one more time I see what was the reasoning.

Reply:
fcab985 Agree on the latter, but I took that the whole point for adding this check in contract was because something bad can happen with the governance, hence the described scenerio context.